rest/route/pod_agent.go

```diff
@@ -443,44 +434,57 @@ func (h *podAgentEndTask) Run(ctx context.Context) gimlet.Responder {
 		})
 	}
 
-	// mark task as finished
 	deactivatePrevious := utility.FromBoolPtr(projectRef.DeactivatePrevious)
 	err = model.MarkEnd(t, evergreen.APIServerTaskActivator, finishTime, &h.details, deactivatePrevious)
 	if err != nil {
 		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "calling mark finish on task %v", t.Id))
 	}
 
+	// Clear the running task on the pod now that the task has finished.
+	if err = p.ClearRunningTask(); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "clearing running task '%s' for pod '%s'", t.Id, p.ID))
+	}
+
 	if t.Requester == evergreen.MergeTestRequester {
 		if err = model.HandleEndTaskForCommitQueueTask(t, h.details.Status); err != nil {
-			return gimlet.MakeJSONErrorResponder(err)
+			return gimlet.MakeJSONInternalErrorResponder(err)
 		}
 	}
 
 	// the task was aborted if it is still in undispatched.
 	// the active state should be inactive.
 	if h.details.Status == evergreen.TaskUndispatched {
 		if t.Activated {
-			grip.Warningf("task %v is active and undispatched after being marked as finished", t.Id)
+			grip.Warningf("task %s is active and undispatched after being marked as finished", t.Id)
 			return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
 		}
-		grip.Infof(fmt.Sprintf("task %v has been aborted and will not run", t.Id))
+		grip.Info(message.Fields{
+			"message": fmt.Sprintf("task %s has been aborted", t.Id),
+			"task_id": t.Id,
+			"path":    fmt.Sprintf("/rest/v2/pods/%s/task/%s/end", p.ID, t.Id),
+		})
 		return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
 	}
 
 	// GetDisplayTask will set the DisplayTask on t if applicable
-	// we set this before the collect task end job is run to prevent data race
 	dt, err := t.GetDisplayTask()
 	if err != nil {
 		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "getting display task"))
 	}
 	queue := h.env.RemoteQueue()
-	job := units.NewCollectTaskEndDataJob(t, nil, p)
+	job := units.NewCollectTaskEndDataJob(t, nil, p, p.ID)
 	if err = queue.Put(ctx, job); err != nil {
-		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "couldn't queue job to update task stats accounting"))
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "couldn't queue job to update task stats accounting"))
 	}
 
 	if p.Status != pod.StatusRunning {
-		endTaskResp.ShouldExit = true
+		j := units.NewPodTerminationJob(h.podID, "pod is no longer running", utility.RoundPartOfMinute(0))
+		if err := amboy.EnqueueUniqueJob(ctx, h.env.RemoteQueue(), j); err != nil {
+			return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
```

Comment:
nit: `gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "enqueueing job to terminate pod"))`